src/audio/module_adapter/module/cadence.c

```diff
@@ -207,16 +254,100 @@ static int cadence_codec_post_init(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
 static int cadence_codec_init(struct processing_module *mod)
 {
-	int ret;
+	const struct ipc4_cadence_module_cfg *cfg;
+	struct module_data *codec = &mod->priv;
+	struct cadence_codec_data *cd;
+	struct module_config *setup_cfg;
 	struct comp_dev *dev = mod->dev;
+	int ret;
+
+	comp_dbg(dev, "cadence_codec_init() start");
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
+	if (!cd) {
+		comp_err(dev, "cadence_codec_init(): failed to allocate memory for cadence codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+	codec->mpd.init_done = 0;
+	cd->self = NULL;
+	cd->mem_tabs = NULL;
+	cd->api = NULL;
+	cd->setup_cfg.avail = false;
+
+	/* copy the setup config only for the first init */
+	if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
+		setup_cfg = &cd->setup_cfg;
+
+		cfg = (const struct ipc4_cadence_module_cfg *)codec->cfg.init_data;
+
+		/* allocate memory for set up config */
+		setup_cfg->data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  cfg->param_size);
+		if (!setup_cfg->data) {
+			comp_err(dev, "cadence_codec_init(): failed to alloc setup config");
+			ret = -ENOMEM;
+			goto free;
+		}
+
+		/* allocate memory for runtime set up config */
+		codec->cfg.data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  cfg->param_size);
+		if (!codec->cfg.data) {
+			comp_err(dev, "cadence_codec_init(): failed to alloc runtime setup config");
+			ret = -ENOMEM;
+			goto free_cfg;
+		}
+
+		codec->cfg.size = cfg->param_size;
+		ret = memcpy_s(codec->cfg.data, codec->cfg.size,
+			       cfg->param, cfg->param_size);
+		if (ret) {
+			comp_err(dev, "cadence_codec_init(): failed to init runtime config %d",
+				 ret);
+			goto free_cfg2;
+		}
+		codec->cfg.avail = true;
+
+		setup_cfg->size = cfg->param_size;
+		ret = memcpy_s(setup_cfg->data, setup_cfg->size,
+			       cfg->param, cfg->param_size);
+		if (ret) {
+			comp_err(dev, "cadence_codec_init(): failed to copy setup config %d", ret);
+			goto free_cfg2;
+		}
+		setup_cfg->avail = true;
+	}
+
+	comp_dbg(dev, "cadence_codec_init() done");
+
+	return 0;
+
+free_cfg2:
+	rfree(codec->cfg.data);
+free_cfg:
+	rfree(setup_cfg->data);
+free:
+	rfree(cd);
+	return ret;
+}
+
+#elif CONFIG_IPC_MAJOR_3
+static int cadence_codec_init(struct processing_module *mod)
+{
 	struct module_data *codec = &mod->priv;
-	struct cadence_codec_data *cd = NULL;
+	struct cadence_codec_data *cd;
+	struct comp_dev *dev = mod->dev;
+	struct module_config *setup_cfg;
+	int ret;
 
 	comp_dbg(dev, "cadence_codec_init() start");
 
-	cd = rballoc(0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
```

Comment:
many fields are initialised to 0 below. Either remove those initialisations or use `rmalloc()`